core/base/inc/ROOT/TSequentialExecutor.hxx

```diff
@@ -13,96 +13,114 @@
 
 #include "RConfigure.h"
 
-#include "ROOT/TExecutor.hxx"
+#include "ROOT/TExecutorCRTP.hxx"
+#include "ExecutionPolicy.hxx"
 #include <numeric>
 #include <vector>
 
 namespace ROOT {
 
-   class TSequentialExecutor: public TExecutor<TSequentialExecutor> {
+   class TSequentialExecutor: public TExecutorCRTP<TSequentialExecutor> {
    public:
       explicit TSequentialExecutor(){};
+      explicit TSequentialExecutor(ROOT::Internal::ExecutionPolicy, unsigned): TSequentialExecutor(){};
 
-      TSequentialExecutor(TSequentialExecutor &) = delete;
-      TSequentialExecutor &operator=(TSequentialExecutor &) = delete;
+      TSequentialExecutor(const TSequentialExecutor &) = delete;
+      TSequentialExecutor &operator=(const TSequentialExecutor &) = delete;
 
       template<class F>
       void Foreach(F func, unsigned nTimes);
       template<class F, class INTEGER>
       void Foreach(F func, ROOT::TSeq<INTEGER> args);
-      /// \cond
       template<class F, class T>
       void Foreach(F func, std::initializer_list<T> args);
-      /// \endcond
       template<class F, class T>
       void Foreach(F func, std::vector<T> &args);
+      template<class F, class T>
+      void Foreach(F func, const std::vector<T> &args);
 
-      using TExecutor<TSequentialExecutor>::Map;
+      using TExecutorCRTP<TSequentialExecutor>::Map;
```

Comment:
I find the mixture of overloads that the derived class implements, and convenience overloadds implemented in the base which wrap overloads provided by the derived class very difficult to understand / follow / remember. I think it's also calling for trouble, because some argument type changed somewhere can cause havoc to the expected derived / base distribution / reimplementation of signatures.

I.e. is `auto Map(F func, std::vector<T> &args)` expected to be implemented in the base, derived, or both?

Maybe rename that single overload (?) that derived classes are expected to provide `MapImpl`, and do not declare it in TExecutorCRTP (or `= delete`)?

Reply:
It grew more confusing over time...

Back in the day, we programmed this class with the idea that it could be used as the generalized executor, by changing the class it was templated on: TExecutorCRTP<TThreadExecutor>, TExecutorCRTP<TProcessExecutor>, etc.

That's why all the Derived() calls and why it doesn't have any deleted function. Deleting a function would break this. But I completely missed the potential infinite recursions, so... I will rethink this